src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -119,66 +119,64 @@ private Task<Socket> AcceptAsyncApm(Socket acceptSocket)
 
         internal Task ConnectAsync(EndPoint remoteEP)
         {
-            var tcs = new TaskCompletionSource<bool>(this);
-            BeginConnect(remoteEP, iar =>
+            // Use ValueTaskReceive so the AwaitableSocketAsyncEventArgs can be re-used later.
+            AwaitableSocketAsyncEventArgs saea = LazyInitializer.EnsureInitialized(ref EventArgs.ValueTaskReceive, () => new AwaitableSocketAsyncEventArgs());
+
+            if (!saea.Reserve())
             {
-                var innerTcs = (TaskCompletionSource<bool>)iar.AsyncState;
-                try
-                {
-                    ((Socket)innerTcs.Task.AsyncState).EndConnect(iar);
-                    innerTcs.TrySetResult(true);
-                }
-                catch (Exception e) { innerTcs.TrySetException(e); }
-            }, tcs);
-            return tcs.Task;
+                saea = new AwaitableSocketAsyncEventArgs();
+                saea.Reserve();
+            }
+
+            saea.RemoteEndPoint = remoteEP;
+            return saea.ConnectAsync(this).AsTask();
         }
 
         internal Task ConnectAsync(IPAddress address, int port)
+            => ConnectAsync(new IPEndPoint(address, port));
+
+        internal async Task ConnectAsync(IPAddress[] addresses, int port)
         {
-            var tcs = new TaskCompletionSource<bool>(this);
-            BeginConnect(address, port, iar =>
+            if (addresses == null)
             {
-                var innerTcs = (TaskCompletionSource<bool>)iar.AsyncState;
-                try
-                {
-                    ((Socket)innerTcs.Task.AsyncState).EndConnect(iar);
-                    innerTcs.TrySetResult(true);
-                }
-                catch (Exception e) { innerTcs.TrySetException(e); }
-            }, tcs);
-            return tcs.Task;
-        }
+                throw new ArgumentNullException(nameof(addresses));
+            }
+            if (addresses.Length == 0)
+            {
+                throw new ArgumentException(SR.net_invalidAddressList, nameof(addresses));
+            }
```

Comment:
As this is an async method, these argument exceptions are going to be stored in the task returned from ConnectAsync.  They should instead be thrown from a non-async method that then delegates to this one.

Reply:
Is this request performance related?

Note that some exceptions that were synchronous before are now requiring an await (because they are happening from Socket.ConnectAsync(saea), instead of BeginConnect).

Should I put this in an async local function? Or a separate function?

Reply:
> Is this request performance related?

No, just expectations and to some extent reliability. We generally want usage exceptions to emerge synchronously. This is particularly  useful if the returned task is consumed in a way that would prevent the exception from being seen, e.g. if it was in a wait all with another operation that required the first in order to complete, resulting in deadlock that's harder to diagnose.

> Should I put this in an async local function? Or a separate function?

In my experience, async local instance methods have more overhead (it's possible IL has improved since I last looked). So for now, a separate method.